File: src/client/java/com/mod/client/KeyMakerFeature.java
package com.mod.client;

import com.mod.client.compat.ContainerCompat;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.screens.inventory.AbstractContainerScreen;
import net.minecraft.network.chat.Component;
import net.minecraft.world.inventory.ChestMenu;
import net.minecraft.world.inventory.Slot;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.Items;

import java.util.List;

public class KeyMakerFeature {

    private KeyMakerState state = KeyMakerState.IDLE;
    private CurrentCraftType currentType = CurrentCraftType.TUNGSTEN;
    private boolean bothSwitched = false;
    private int currentSlotIndex = 0;
    private int actionSlot = -1;
    private long lastActionTime = 0;

    public void tick(Minecraft client) {
        if (!OneAddons.keyMakerEnabled) {
            reset();
            return;
        }

        if (client.player == null || client.level == null || client.gameMode == null) {
            reset();
            return;
        }

        if (OneAddons.keyMakerMode != KeyMode.BOTH) {
            currentType = OneAddons.keyMakerMode == KeyMode.UMBER
                ? CurrentCraftType.UMBER
                : CurrentCraftType.TUNGSTEN;
        }

        switch (state) {
            case IDLE -> tickIdle(client);
            case SCANNING -> tickScanning(client);
            case COLLECTING -> tickCollecting(client);
            case CRAFT_FURNACE -> tickCraftFurnace(client);
            case WAIT_PROCESS -> tickWaitProcess(client);
            case CONFIRM -> tickConfirm(client);
            case CLOSE_ABORT -> tickCloseAbort(client);
            case WAIT_FORGE_RETURN -> tickWaitForgeReturn(client);
        }
    }

    private void reset() {
        state = KeyMakerState.IDLE;
        currentType = OneAddons.keyMakerMode == KeyMode.UMBER
            ? CurrentCraftType.UMBER
            : CurrentCraftType.TUNGSTEN;
        bothSwitched = false;
        currentSlotIndex = 0;
    }

    private void tickIdle(Minecraft client) {
        if (isForgeOpen(client)) {
            state = KeyMakerState.SCANNING;
            currentSlotIndex = 0;
        }
    }

    private void tickScanning(Minecraft client) {
        if (!isForgeOpen(client)) {
            reset();
            return;
        }

        ForgeSlot[] slots = SlotMappings.FORGE_PROCESS_SLOTS;

        for (int i = 0; i < slots.length; i++) {
            int idx = (currentSlotIndex + i) % slots.length;
            ForgeSlot slot = slots[idx];
            if (isSlotCompleted(client, slot)) {
                currentSlotIndex = (idx + 1) % slots.length;
                actionSlot = slot.topSlot();
                state = KeyMakerState.COLLECTING;
                return;
            }
        }

        if (!readyToClick()) return;

        for (int i = 0; i < slots.length; i++) {
            int idx = (currentSlotIndex + i) % slots.length;
            ForgeSlot slot = slots[idx];
            if (isSlotEmpty(client, slot)) {
                currentSlotIndex = (idx + 1) % slots.length;
                actionSlot = slot.topSlot();
                performClick(client, actionSlot);
                state = KeyMakerState.CRAFT_FURNACE;
                return;
            }
        }

        currentSlotIndex = (currentSlotIndex + 1) % slots.length;
    }

    private void tickCollecting(Minecraft client) {
        if (!isForgeOpen(client)) {
            return;
        }

        if (!readyToClick()) return;

        ForgeSlot forgeSlot = findSlotByTop(actionSlot);
        if (forgeSlot == null || !isSlotCompleted(client, forgeSlot)) {
            state = KeyMakerState.SCANNING;
            return;
        }

        performClick(client, actionSlot);
        state = KeyMakerState.SCANNING;
    }

    private void tickCraftFurnace(Minecraft client) {
        if (client.screen == null) {
            return;
        }

        if (!readyToClick()) return;

        AbstractContainerScreen<?> screen = getScreen(client);
        if (screen == null) {
            return;
        }

        String title = screen.getTitle().getString();
        if (title.contains(SlotMappings.SELECT_PROCESS_TITLE)) {
            performClick(client, SlotMappings.SELECT_PROCESS_OTHER_SLOT);
            state = KeyMakerState.WAIT_PROCESS;
        } else if (title.contains(SlotMappings.FORGE_TITLE)) {
            return;
        } else {
            reset();
        }
    }

    private void tickWaitProcess(Minecraft client) {
        if (client.screen == null) return;
        if (!readyToClick()) return;

        int keySlot = currentType == CurrentCraftType.TUNGSTEN
            ? SlotMappings.KEY_SELECTION_TUNGSTEN_SLOT
            : SlotMappings.KEY_SELECTION_UMBER_SLOT;
        performClick(client, keySlot);
        state = KeyMakerState.CONFIRM;
    }

    private void tickConfirm(Minecraft client) {
        if (client.screen == null) return;
        if (!readyToClick()) return;

        AbstractContainerScreen<?> screen = getScreen(client);
        if (screen == null) return;

        String title = screen.getTitle().getString();
        if (title.contains(SlotMappings.CONFIRM_PROCESS_TITLE) || title.contains("Confirm")) {
            performClick(client, SlotMappings.CONFIRM_BUTTON_SLOT);
            state = KeyMakerState.WAIT_FORGE_RETURN;
        } else if (title.contains(SlotMappings.FORGE_TITLE)) {
            state = KeyMakerState.SCANNING;
        } else if (title.contains(SlotMappings.SELECT_PROCESS_TITLE)) {
            return;
        } else {
            reset();
        }
    }

    private void tickWaitForgeReturn(Minecraft client) {
        if (isForgeOpen(client)) {
            state = KeyMakerState.SCANNING;
            currentSlotIndex = 0;
        } else if (readyToClick() && isConfirmOpen(client)) {
            handleExhausted(client);
        }
    }

    private void tickCloseAbort(Minecraft client) {
        if (!readyToClick()) return;

        if (client.player != null) {
            client.player.closeContainer();
        }

        KeyMode mode = OneAddons.keyMakerMode;

        if (mode == KeyMode.BOTH && currentType == CurrentCraftType.TUNGSTEN) {
            currentType = CurrentCraftType.UMBER;
            bothSwitched = true;
            state = KeyMakerState.WAIT_FORGE_RETURN;
        } else {
            sendNotification(client);
            reset();
        }
    }

    private void handleExhausted(Minecraft client) {
        state = KeyMakerState.CLOSE_ABORT;
        lastActionTime = System.currentTimeMillis();
    }

    private void sendNotification(Minecraft client) {
        if (client.player == null) return;

        KeyMode mode = OneAddons.keyMakerMode;
        String msg;

        if (mode == KeyMode.BOTH) {
            msg = "[KeyMaker] Finished crafting all available Tungsten and Umber Keys.";
        } else if (mode == KeyMode.TUNGSTEN) {
            msg = "[KeyMaker] Finished crafting Tungsten Keys.";
        } else {
            msg = "[KeyMaker] Finished crafting Umber Keys.";
        }

        client.player.sendSystemMessage(Component.literal(msg));
    }

    private boolean isForgeOpen(Minecraft client) {
        AbstractContainerScreen<?> screen = getScreen(client);
        if (screen == null) return false;
        String title = screen.getTitle().getString();
        return title.contains(SlotMappings.FORGE_TITLE);
    }

    private boolean isSlotEmpty(Minecraft client, ForgeSlot slot) {
        ChestMenu menu = getMenu(client);
        if (menu == null) return false;

        List<Slot> slots = menu.slots;
        if (slot.topSlot() >= slots.size() || slot.lastGlassSlot() >= slots.size()) return false;

        ItemStack top = slots.get(slot.topSlot()).getItem();
        ItemStack lastGlass = slots.get(slot.lastGlassSlot()).getItem();

        return SlotMappings.isFurnace(top) && lastGlass.getItem() == Items.RED_STAINED_GLASS_PANE;
    }

    private boolean isSlotCompleted(Minecraft client, ForgeSlot slot) {
        ChestMenu menu = getMenu(client);
        if (menu == null) return false;

        List<Slot> slots = menu.slots;
        if (slot.topSlot() >= slots.size() || slot.lastGlassSlot() >= slots.size()) return false;

        ItemStack lastGlass = slots.get(slot.lastGlassSlot()).getItem();
        ItemStack top = slots.get(slot.topSlot()).getItem();
        return lastGlass.getItem() == Items.LIME_STAINED_GLASS_PANE && !top.isEmpty() && !SlotMappings.isFurnace(top);
    }

    private ForgeSlot findSlotByTop(int topSlot) {
        for (ForgeSlot slot : SlotMappings.FORGE_PROCESS_SLOTS) {
            if (slot.topSlot() == topSlot) return slot;
        }
        return null;
    }

    private boolean isConfirmOpen(Minecraft client) {
        AbstractContainerScreen<?> screen = getScreen(client);
        if (screen == null) return false;
        String title = screen.getTitle().getString();
        return title.contains(SlotMappings.CONFIRM_PROCESS_TITLE) || title.contains("Confirm");
    }

    private boolean readyToClick() {
        long delay = OneAddons.keyMakerClickDelay;
        return System.currentTimeMillis() - lastActionTime >= delay;
    }

    private void performClick(Minecraft client, int slotIndex) {
        ChestMenu menu = getMenu(client);
        if (menu == null) return;
        ContainerCompat.leftClick(menu.containerId, slotIndex);
        lastActionTime = System.currentTimeMillis();
    }

    private AbstractContainerScreen<?> getScreen(Minecraft client) {
        if (client.screen instanceof AbstractContainerScreen<?> screen) {
            return screen;
        }
        return null;
    }

    private ChestMenu getMenu(Minecraft client) {
        AbstractContainerScreen<?> screen = getScreen(client);
        if (screen != null && screen.getMenu() instanceof ChestMenu menu) {
            return menu;
        }
        return null;
    }
}

File: src/client/kotlin/com/mod/client/KeyMakerFeature.kt
package com.mod.client

import com.mod.client.compat.ContainerCompat
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.screens.inventory.AbstractContainerScreen
import net.minecraft.network.chat.Component
import net.minecraft.world.inventory.ChestMenu
import net.minecraft.world.item.Items

class KeyMakerFeature {

    private var state = KeyMakerState.IDLE
    private var currentType = CurrentCraftType.TUNGSTEN
    private var currentSlotIndex = 0
    private var actionSlot = -1
    private var lastActionTime = 0L

    fun tick(client: Minecraft) {
        if (!OneAddons.keyMakerEnabled) {
            reset()
            return
        }

        if (client.player == null || client.level == null || client.gameMode == null) {
            reset()
            return
        }

        currentType = if (OneAddons.keyMakerMode == KeyMode.UMBER)
            CurrentCraftType.UMBER
        else
            CurrentCraftType.TUNGSTEN

        when (state) {
            KeyMakerState.IDLE -> tickIdle(client)
            KeyMakerState.SCANNING -> tickScanning(client)
            KeyMakerState.COLLECTING -> tickCollecting(client)
            KeyMakerState.CRAFT_FURNACE -> tickCraftFurnace(client)
            KeyMakerState.WAIT_PROCESS -> tickWaitProcess(client)
            KeyMakerState.CONFIRM -> tickConfirm(client)
            KeyMakerState.CLOSE_ABORT -> tickCloseAbort(client)
            KeyMakerState.WAIT_FORGE_RETURN -> tickWaitForgeReturn(client)
        }
    }

    private fun reset() {
        state = KeyMakerState.IDLE
        currentType = if (OneAddons.keyMakerMode == KeyMode.UMBER)
            CurrentCraftType.UMBER
        else
            CurrentCraftType.TUNGSTEN
        currentSlotIndex = 0
    }

    private fun tickIdle(client: Minecraft) {
        if (isForgeOpen(client)) {
            state = KeyMakerState.SCANNING
            currentSlotIndex = 0
        }
    }

    private fun tickScanning(client: Minecraft) {
        if (!isForgeOpen(client)) {
            reset()
            return
        }

        val slots = SlotMappings.FORGE_PROCESS_SLOTS

        for (i in slots.indices) {
            val idx = (currentSlotIndex + i) % slots.size
            val slot = slots[idx]
            if (isSlotCompleted(client, slot)) {
                currentSlotIndex = (idx + 1) % slots.size
                actionSlot = slot.topSlot
                state = KeyMakerState.COLLECTING
                return
            }
        }

        if (!readyToClick()) return

        for (i in slots.indices) {
            val idx = (currentSlotIndex + i) % slots.size
            val slot = slots[idx]
            if (isSlotEmpty(client, slot)) {
                currentSlotIndex = (idx + 1) % slots.size
                actionSlot = slot.topSlot
                performClick(client, actionSlot)
                state = KeyMakerState.CRAFT_FURNACE
                return
            }
        }

        currentSlotIndex = (currentSlotIndex + 1) % slots.size
    }

    private fun tickCollecting(client: Minecraft) {
        if (!isForgeOpen(client)) {
            return
        }

        if (!readyToClick()) return

        val forgeSlot = findSlotByTop(actionSlot)
        if (forgeSlot == null || !isSlotCompleted(client, forgeSlot)) {
            state = KeyMakerState.SCANNING
            return
        }

        performClick(client, actionSlot)
        state = KeyMakerState.SCANNING
    }

    private fun tickCraftFurnace(client: Minecraft) {
        if (client.screen == null) {
            return
        }

        if (!readyToClick()) return

        val screen = getScreen(client) ?: return

        val title = screen.title.string
        if (title.contains(SlotMappings.SELECT_PROCESS_TITLE)) {
            performClick(client, SlotMappings.SELECT_PROCESS_OTHER_SLOT)
            state = KeyMakerState.WAIT_PROCESS
        } else if (title.contains(SlotMappings.FORGE_TITLE)) {
            return
        } else {
            reset()
        }
    }

    private fun tickWaitProcess(client: Minecraft) {
        if (client.screen == null) return
        if (!readyToClick()) return

        val keySlot = if (currentType == CurrentCraftType.TUNGSTEN)
            SlotMappings.KEY_SELECTION_TUNGSTEN_SLOT
        else
            SlotMappings.KEY_SELECTION_UMBER_SLOT
        performClick(client, keySlot)
        state = KeyMakerState.CONFIRM
    }

    private fun tickConfirm(client: Minecraft) {
        if (client.screen == null) return
        if (!readyToClick()) return

        val screen = getScreen(client) ?: return

        val title = screen.title.string
        if (title.contains(SlotMappings.CONFIRM_PROCESS_TITLE) || title.contains("Confirm")) {
            performClick(client, SlotMappings.CONFIRM_BUTTON_SLOT)
            state = KeyMakerState.WAIT_FORGE_RETURN
        } else if (title.contains(SlotMappings.FORGE_TITLE)) {
            state = KeyMakerState.SCANNING
        } else if (title.contains(SlotMappings.SELECT_PROCESS_TITLE)) {
            return
        } else {
            reset()
        }
    }

    private fun tickWaitForgeReturn(client: Minecraft) {
        if (isForgeOpen(client)) {
            state = KeyMakerState.SCANNING
            currentSlotIndex = 0
        } else if (readyToClick() && isConfirmOpen(client)) {
            handleExhausted(client)
        }
    }

    private fun tickCloseAbort(client: Minecraft) {
        if (!readyToClick()) return

        if (client.player != null) {
            client.player!!.closeContainer()
        }

        sendNotification(client)
        reset()
    }

    private fun handleExhausted(client: Minecraft) {
        state = KeyMakerState.CLOSE_ABORT
        lastActionTime = System.currentTimeMillis()
    }

    private fun sendNotification(client: Minecraft) {
        val player = client.player ?: return

        val msg = if (OneAddons.keyMakerMode == KeyMode.TUNGSTEN)
            "[KeyMaker] Finished crafting Tungsten Keys."
        else
            "[KeyMaker] Finished crafting Umber Keys."

        player.sendSystemMessage(Component.literal(msg))
    }

    private fun isForgeOpen(client: Minecraft): Boolean {
        val screen = getScreen(client) ?: return false
        val title = screen.title.string
        return title.contains(SlotMappings.FORGE_TITLE)
    }

    private fun isSlotEmpty(client: Minecraft, slot: ForgeSlot): Boolean {
        val menu = getMenu(client) ?: return false

        val slots = menu.slots
        if (slot.topSlot >= slots.size || slot.lastGlassSlot >= slots.size) return false

        val top = slots[slot.topSlot].item
        val lastGlass = slots[slot.lastGlassSlot].item

        return SlotMappings.isFurnace(top) && lastGlass.item === Items.RED_STAINED_GLASS_PANE
    }

    private fun isSlotCompleted(client: Minecraft, slot: ForgeSlot): Boolean {
        val menu = getMenu(client) ?: return false

        val slots = menu.slots
        if (slot.topSlot >= slots.size || slot.lastGlassSlot >= slots.size) return false

        val lastGlass = slots[slot.lastGlassSlot].item
        val top = slots[slot.topSlot].item
        return lastGlass.item === Items.LIME_STAINED_GLASS_PANE && !top.isEmpty && !SlotMappings.isFurnace(top)
    }

    private fun findSlotByTop(topSlot: Int): ForgeSlot? {
        for (slot in SlotMappings.FORGE_PROCESS_SLOTS) {
            if (slot.topSlot == topSlot) return slot
        }
        return null
    }

    private fun isConfirmOpen(client: Minecraft): Boolean {
        val screen = getScreen(client) ?: return false
        val title = screen.title.string
        return title.contains(SlotMappings.CONFIRM_PROCESS_TITLE) || title.contains("Confirm")
    }

    private fun readyToClick(): Boolean {
        val delay = Math.max(500, OneAddons.keyMakerClickDelay)
        return System.currentTimeMillis() - lastActionTime >= delay
    }

    private fun performClick(client: Minecraft, slotIndex: Int) {
        val menu = getMenu(client) ?: return
        ContainerCompat.leftClick(menu.containerId, slotIndex)
        lastActionTime = System.currentTimeMillis()
    }

    private fun getScreen(client: Minecraft): AbstractContainerScreen<*>? {
        return client.screen as? AbstractContainerScreen<*>
    }

    private fun getMenu(client: Minecraft): ChestMenu? {
        val screen = getScreen(client) ?: return null
        return screen.menu as? ChestMenu
    }
}
